src/cr_kyoushi/generator/template.py

```diff
@@ -52,9 +52,165 @@
     )
 
 
+def _normalize_propabilities(
+    propabilities: Sequence[float],
+    check_positive: bool = True,
+) -> List[float]:
+    """Normalizes a propability distribution to sum up to 1.
+
+    Has a floating point error tolerance of up to 1e-8 (same as numpy.choice).
+    If the resulting difference is greater than 1e-8 it is added to the first
+    non zero propability.
+
+    Args:
+        propabilities: The distribution to normalize
+        check_positive: Switch to disable the positive number test.
+                        This makes the function slightly faster
+
+    Raises:
+        ValueError: If the distribution sums to 0 or
+                    If one of the propability values is negative
+
+    Returns:
+        The normalized probability distribution.
+    """
+    # only check when requested
+    if check_positive and any(p < 0 for p in propabilities):
+        raise AnsibleFilterError(
+            f"Propabilities must be positive numbers, but got {propabilities}"
+        )
+
+    total = sum(propabilities)
+
+    if abs(1.0 - total) < 1e-8:
+        # if we already have a normalized
+        # distribution then we can just return it
+        return propabilities
+    elif total > 0:
+        multiplier = 1.0 / total
+        propabilities = [p * multiplier for p in propabilities]
+        diff = 1.0 - sum(propabilities)
+        # when rounding errors become to extrem
+        # then we fix them by adding the diff to 1 a propability
+        if abs(diff) > 1e-8:
+            # first non 0 propability index will be increased
+            adjust_index = next((i for i, x in enumerate(propabilities) if x != 0.0))
+            propabilities[adjust_index] += diff
+        return propabilities
+
+    raise Exception(f"Propabilities sum to 0, but got {propabilities}")
+
+
+def normalize_propabilities(propabilities, ignore=["extra"]):
+    """Accepts probability distribution dicts and lists and normalizes them.
+
+    Args:
+        propabilities (Union[Dict, List]): A probability distribution as dict or list
+        ignore (List[str], optional): Additional dict keys to ignore during normalization.
+
+    Raises:
+        AnsibleFilterError: If the given distribution has an invalid container type, sums to 0
```

Comment:
Also need to update all the doc strings to reflect that this does not use Ansible Errors here